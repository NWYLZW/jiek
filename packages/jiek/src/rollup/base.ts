import type { InputPluginOption, OutputOptions } from 'rollup'

export type Mapping2ROO<K extends keyof OutputOptions> = OutputOptions[K] | {
  js?: OutputOptions[K]
  dts?: OutputOptions[K]
}

export interface ConfigGenerateContext {
  path: string
  name: string
  input: string
  output: string
  external: (string | RegExp)[]
  pkgIsModule: boolean
  conditionals: string[]
}

export type OutputControl = boolean | ((context: ConfigGenerateContext) => boolean)

export const BUILDER_TYPES = ['esbuild', 'swc'] as const

export const BUILDER_TYPE_PACKAGE_NAME_MAP = {
  esbuild: 'rollup-plugin-esbuild',
  swc: 'rollup-plugin-swc3'
}

export interface TemplateOptions {
  /**
   * When the user configures type: module, the generated output from entry points that don't
   * have cts as a suffix will automatically include the CJS version.
   * if it is not configured, and the generated output from entry points that do not have mts
   * as a suffix will automatically include the ESM version.
   *
   * @default true
   */
  crossModuleConvertor?: boolean
  /**
   * Auto-detect the builder from the installed dependencies.
   * If the builder is not installed, it will prompt the user to install it.
   * If exists multiple builders, it will fall back to the 'esbuild'.
   *
   * @default 'esbuild'
   */
  builder?:
    | typeof BUILDER_TYPES[number]
    | ({
      type: 'esbuild'
    } & import('rollup-plugin-esbuild').Options)
    | ({
      type: 'swc'
    } & import('rollup-plugin-swc3').PluginOptions)
  features?: {
    /**
     * When use esbuild type builder, it will inject `supported.import-attributes` option.
     * When use swc type builder, it will inject `jsc.experimental.keepImportAttributes` option.
     *
     * And it will auto set the rollup output `externalImportAttributes` and `importAttributesKey` options.
     *
     * @default false
     */
    keepImportAttributes?: boolean | 'assert'
  }
  output?: {
    /**
     * @default true
     *
     * When minify is set to true, the output will with minified files.
     * When minify is set to 'only-minify', the output will direct output minified files.
     */
    minify?: boolean | 'only-minify'
    minifyOptions?:
      | typeof BUILDER_TYPES[number]
      | 'terser'
      | (
        {
          type: 'terser'
        } & import('@rollup/plugin-terser').Options
      )
      | (
        {
          type: 'esbuild'
        } & Parameters<typeof import('rollup-plugin-esbuild').minify>[0]
      )
      | (
        {
          type: 'swc'
        } & Parameters<typeof import('rollup-plugin-swc3').minify>[0]
      )
    /**
     * @default 'dist'
     */
    dir?: Mapping2ROO<'dir'>
    sourcemap?: Mapping2ROO<'sourcemap'>
    strict?: Mapping2ROO<'strict'>
    js?: OutputControl
    dts?: OutputControl
  }
  /**
   * Set the external dependencies of the package.
   */
  external?: (string | RegExp)[]
  plugins?:
    | InputPluginOption
    | ((type: 'js' | 'dts', context: ConfigGenerateContext) => InputPluginOption)
    | {
      js: InputPluginOption
      dts?: InputPluginOption
    }
    | {
      js?: InputPluginOption
      dts: InputPluginOption
    }
  /**
   * https://www.npmjs.com/package/@rollup/plugin-inject#usage
   *
   * @example
   * ```js
   * {
   *   // import { Promise } from 'es6-promise'
   *   Promise: [ 'es6-promise', 'Promise' ],
   *
   *   // import { Promise as P } from 'es6-promise'
   *   P: [ 'es6-promise', 'Promise' ],
   *
   *   // import $ from 'jquery'
   *   $: 'jquery',
   *
   *   // import * as fs from 'fs'
   *   fs: [ 'fs', '*' ],
   *
   *   // use a local module instead of a third-party one
   *   'Object.assign': path.resolve( 'src/helpers/object-assign.js' ),
   * }
   * ```
   */
  injects?: Record<string, string | [string, string]>
}
