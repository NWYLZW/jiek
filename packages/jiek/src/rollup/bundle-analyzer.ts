import process from 'node:process'

import type { InputPluginOption, OutputPlugin, Plugin } from 'rollup'
import type { AnalyzerPluginInternalAPI } from 'vite-bundle-analyzer'

export type Module = ReturnType<AnalyzerPluginInternalAPI['processModule']>[number]

const {
  JIEK_ANALYZER
} = process.env

const ANALYZER = (JIEK_ANALYZER != null) && JSON.parse(JIEK_ANALYZER) as {
  dir?: string
  mode?: string
  size?: string
  port?: number
  open?: boolean
}

export function bundleAnalyzer(modulesResolved: (modules: Module[]) => void) {
  // eslint-disable-next-line ts/strict-boolean-expressions
  if (!ANALYZER) {
    return []
  }

  const defaultSizes = ({
    parsed: 'parsed',
    stat: 'stat',
    gzip: 'gzip'
  } as const)[ANALYZER.size ?? 'stat'] ?? 'parsed'

  let module: typeof import('vite-bundle-analyzer') | undefined
  let ana: Plugin | undefined
  async function initAna() {
    const { adapter, analyzer } = module ?? await import('vite-bundle-analyzer')
    ana = ana ?? adapter(analyzer({
      defaultSizes,
      analyzerMode: modulesResolved
    }))
  }

  return [
    (async () => {
      await initAna()
      return {
        name: 'jiek:bundle-analyzer',
        async closeBundle(...args) {
          if (typeof ana!.closeBundle !== 'function') return

          return ana!.closeBundle?.call(this, ...args)
        }
      } satisfies InputPluginOption
    })(),
    (async () => {
      await initAna()
      return {
        ...ana,
        name: 'jiek:bundle-analyzer-output'
      } satisfies OutputPlugin
    })()
  ] as const
}
