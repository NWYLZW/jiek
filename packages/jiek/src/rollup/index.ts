import fs from 'node:fs'
import { dirname, resolve } from 'node:path'

import type { RecursiveRecord } from '@jiek/pkger/entrypoints'
import {
  DEFAULT_SKIP_VALUES,
  entrypoints2Exports,
  filterLeafs,
  getAllLeafs,
  resolveEntrypoints
} from '@jiek/pkger/entrypoints'
import { dts } from '@jiek/rollup-plugin-dts'
import { getWorkspaceDir } from '@jiek/utils/getWorkspaceDir'
import json from '@rollup/plugin-json'
import { nodeResolve } from '@rollup/plugin-node-resolve'
import terser from '@rollup/plugin-terser'
import { sendMessage } from 'execa'
import { parse } from 'jsonc-parser'
import { isMatch } from 'micromatch'
import type { OutputOptions, OutputPlugin, RollupOptions } from 'rollup'
import esbuild from 'rollup-plugin-esbuild'
import ts from 'typescript'

import type { RollupProgressEvent } from './base'
import progress from './plugins/progress'
import skip from './plugins/skip'
import externalResolver from './utils/externalResolver'

export interface TemplateOptions {
}

interface PackageJSON {
  name?: string
  type?: string
  exports?: Record<string, unknown> | string | string[]
}

const {
  JIEK_ROOT
} = process.env
const WORKSPACE_ROOT = JIEK_ROOT ?? getWorkspaceDir()
const COMMON_OPTIONS = {} satisfies RollupOptions
const COMMON_PLUGINS = [
  json()
]

const STYLE_REGEXP = /\.(css|s[ac]ss|less|styl)$/

// eslint-disable-next-line unused-imports/no-unused-vars
const debug = (...args: unknown[]) => sendMessage({ type: 'debug', data: args } satisfies RollupProgressEvent)

const resolveWorkspacePath = (p: string) => resolve(WORKSPACE_ROOT, p)

const pascalCase = (str: string) =>
  str
    .replace(/[@|/-](\w)/g, (_, $1) => $1.toUpperCase())
    .replace(/(?:^|-)(\w)/g, (_, $1) => $1.toUpperCase())

const reveal = (obj: string | Record<string, unknown>, keys: string[]) =>
  keys.reduce((acc, key) => {
    if (typeof acc === 'string') throw new Error('key not found in exports')
    if (!(key in acc)) throw new Error(`key ${key} not found in exports`)
    return acc[key] as string | Record<string, unknown>
  }, obj)

const withMinify = (
  output: OutputOptions & {
    plugins?: OutputPlugin[]
  }
) => [
  output,
  {
    ...output,
    file: output.file?.replace(/(\.[cm]?js)$/, '.min$1'),
    plugins: [
      ...(output.plugins ?? []),
      terser()
    ]
  }
]

type TSConfig = {
  extends?: string | string[]
  compilerOptions?: Record<string, unknown>
  references?: { path: string }[]
  files?: string[]
  include?: string[]
  exclude?: string[]
}

const getTSConfig = (p: string): TSConfig =>
  !fs.existsSync(p) || !fs.statSync(p).isFile()
    ? {}
    : parse(fs.readFileSync(p, 'utf-8'), [], { allowTrailingComma: true, allowEmptyContent: true })

const getExtendTSConfig = (tsconfigPath: string): TSConfig => {
  tsconfigPath = resolve(tsconfigPath)
  const tsconfigPathDirname = dirname(tsconfigPath)
  const { extends: exts, ...tsconfig } = getTSConfig(tsconfigPath)
  const resolvePaths = (paths: string[] | undefined) => paths?.map(p => resolve(tsconfigPathDirname, p)) ?? []

  const extendsPaths = resolvePaths(
    exts ? Array.isArray(exts) ? exts : [exts] : []
  )
  if (extendsPaths.length === 0) return tsconfig
  return extendsPaths
    .map(getExtendTSConfig)
    // https://www.typescriptlang.org/tsconfig/#files:~:text=Currently%2C%20the%20only%20top%2Dlevel%20property%20that%20is%20excluded%20from%20inheritance%20is%20references.
    // Currently, the only top-level property that is excluded from inheritance is references.
    .reduce((acc, { compilerOptions = {}, references: _, ...curr }) => ({
      ...acc,
      ...curr,
      compilerOptions: {
        ...acc.compilerOptions,
        ...compilerOptions
      }
    }), tsconfig)
}

const getCompilerOptionsByFilePath = (tsconfigPath: string, filePath: string): Record<string, unknown> | undefined => {
  tsconfigPath = resolve(tsconfigPath)
  filePath = resolve(filePath)
  const tsconfigPathDirname = dirname(tsconfigPath)
  // https://www.typescriptlang.org/tsconfig/#files:~:text=It%E2%80%99s%20worth%20noting%20that%20files%2C%20include%2C%20and%20exclude%20from%20the%20inheriting%20config%20file%20overwrite%20those%20from%20the%20base%20config%20file%2C%20and%20that%20circularity%20between%20configuration%20files%20is%20not%20allowed.
  // Itâ€™s worth noting that files, include, and exclude from the inheriting config file overwrite
  // those from the base config file, and that circularity between configuration files is not allowed.
  const tsconfig = getExtendTSConfig(tsconfigPath)

  const resolvePaths = (paths: string[] | undefined) => paths?.map(p => resolve(tsconfigPathDirname, p)) ?? []

  const [
    references,
    files,
    include,
    exclude
  ] = [
    tsconfig.references?.map(({ path }) => path),
    tsconfig.files,
    tsconfig.include,
    tsconfig.exclude
  ].map(resolvePaths)
  if (exclude.length > 0 && exclude.some(i => isMatch(filePath, i))) return

  // when files or include is not empty, the tsconfig should be ignored
  if (tsconfig.files?.length === 0 && tsconfig.include?.length === 0) return
  let isInclude = false
  isInclude ||= files.length > 0 && files.includes(filePath)
  isInclude ||= include.length > 0 && include.some(i => isMatch(filePath, i))
  if (isInclude) {
    return tsconfig.compilerOptions ?? {}
  } else {
    // when files or include is not empty, but the file is not matched, the tsconfig should be ignored
    if (
      (tsconfig.files && tsconfig.files.length > 0)
      || (tsconfig.include && tsconfig.include.length > 0)
    ) return
  }

  references.reverse()
  for (const ref of references) {
    const compilerOptions = getCompilerOptionsByFilePath(ref, filePath)
    if (compilerOptions) return compilerOptions
  }
  return tsconfig.compilerOptions
}

const generateConfigs = ({
  path,
  name,
  input,
  output,
  outdir,
  external,
  pkgIsModule,
  conditionals
}: {
  path: string
  name: string
  input: string
  output: string
  outdir: string
  external: (string | RegExp)[]
  pkgIsModule: boolean
  conditionals: string[]
}): RollupOptions[] => {
  const isModule = conditionals.includes('import')
  const isCommonJS = conditionals.includes('require')
  const isBrowser = conditionals.includes('browser')
  const dtsTSConfigPaths = [
    resolveWorkspacePath('tsconfig.json'),
    resolveWorkspacePath('tsconfig.dts.json')
  ]
  let dtsTSConfigPath: string | undefined
  dtsTSConfigPaths.forEach(p => {
    if (fs.existsSync(p) && fs.statSync(p).isFile()) {
      dtsTSConfigPath = p
    }
  })
  let compilerOptions: ts.CompilerOptions = {}
  if (dtsTSConfigPath) {
    const jsonCompilerOptions = getCompilerOptionsByFilePath(dtsTSConfigPath, resolve(input))
    const { options, errors } = ts.convertCompilerOptionsFromJson(
      jsonCompilerOptions,
      dirname(dtsTSConfigPath)
    )
    if (errors.length > 0) {
      throw new Error(errors.map(e => e.messageText).join('\n'))
    }
    compilerOptions = options
  }
  const exportConditions = [...conditionals, ...(compilerOptions.customConditions ?? [])]
  const throughEventProps: RollupProgressEvent & { type: 'progress' } = {
    type: 'progress',
    data: { name, path, exportConditions, input }
  }
  return [
    {
      input,
      external,
      output: [
        ...withMinify({
          file: output,
          name,
          format: isModule ? 'esm' : (
            isCommonJS ? 'cjs' : (
              isBrowser ? 'umd' : (
                pkgIsModule ? 'esm' : 'cjs'
              )
            )
          )
        })
      ],
      plugins: [
        nodeResolve({ exportConditions }),
        import('rollup-plugin-postcss')
          .then(({ default: postcss }) =>
            postcss({
              extract: resolve(output.replace(/\.[cm]?js$/, '.css')),
              minimize: true
            })
          )
          .catch(() => void 0),
        esbuild(),
        progress({
          onEvent: (event, message) =>
            sendMessage(
              {
                ...throughEventProps,
                data: { ...throughEventProps.data, event, message, tags: ['js'] }
              } satisfies RollupProgressEvent
            )
        })
      ]
    },
    {
      input,
      external,
      output: [
        {
          dir: outdir,
          entryFileNames: () => input.replace(/^\.\/src\//, '').replace(/(.[cm]?ts)$/, '.d$1')
        }
      ],
      plugins: [
        nodeResolve({ exportConditions }),
        skip({ patterns: [STYLE_REGEXP] }),
        dts({
          respectExternal: true,
          compilerOptions
        }),
        progress({
          onEvent: (event, message) =>
            sendMessage(
              {
                ...throughEventProps,
                data: { ...throughEventProps.data, event, message, tags: ['dts'] }
              } satisfies RollupProgressEvent
            )
        })
      ]
    }
  ]
}

export function template(packageJSON: PackageJSON, options: TemplateOptions = {}) {
  const { name, type, exports: entrypoints } = packageJSON
  const pkgIsModule = type === 'module'
  const outdir = 'dist'
  if (!name) throw new Error('package.json name is required')
  if (!entrypoints) throw new Error('package.json exports is required')

  const packageName = pascalCase(name)

  const external = externalResolver(packageJSON as Record<string, unknown>)

  const [, resolvedEntrypoints] = resolveEntrypoints(entrypoints)
  const filteredResolvedEntrypoints = filterLeafs(
    resolvedEntrypoints as RecursiveRecord<string>,
    {
      skipValue: [
        // ignore values that filename starts with `.jk-noentry`
        /(^|\/)\.jk-noentry/,
        ...DEFAULT_SKIP_VALUES
      ]
    }
  )
  const exports = entrypoints2Exports(filteredResolvedEntrypoints, {})

  const leafMap = new Map<string, string[][]>()
  getAllLeafs(filteredResolvedEntrypoints as RecursiveRecord<string>, ({ keys, value }) => {
    if (typeof value === 'string') {
      const keysArr = leafMap.get(value) ?? []
      leafMap.set(value, keysArr)
      keysArr.push(keys)
    }
    return false
  })

  const configs: RollupOptions[] = []
  leafMap.forEach((keysArr, input) =>
    keysArr.forEach((keys) => {
      const [path, ...conditionals] = keys

      const name = packageName + (path === '.' ? '' : pascalCase(path))
      const keyExports = reveal(exports, keys)
      const commonOptions = {
        path,
        name,
        input,
        outdir,
        external,
        pkgIsModule
      }

      switch (typeof keyExports) {
        case 'string': {
          configs.push(...generateConfigs({
            ...commonOptions,
            output: keyExports,
            conditionals
          }))
          break
        }
        case 'object': {
          getAllLeafs(keyExports as RecursiveRecord<string>, ({ keys: nextKeys, value }) => {
            const allConditionals = [...new Set([...conditionals, ...nextKeys])]
            if (typeof value === 'string') {
              configs.push(...generateConfigs({
                ...commonOptions,
                output: value,
                conditionals: allConditionals
              }))
            }
            return false
          })
          break
        }
      }
    })
  )
  sendMessage(
    {
      type: 'init',
      data: {
        leafMap,
        targetsLength: configs.length
      }
    } satisfies RollupProgressEvent
  )
  return configs.map(c => ({
    ...COMMON_OPTIONS,
    ...c,
    plugins: [
      ...COMMON_PLUGINS,
      c.plugins
    ]
  }))
}
