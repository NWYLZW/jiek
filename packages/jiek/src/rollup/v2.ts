import { resolve } from 'node:path'

import type { RecursiveRecord } from '@jiek/pkger/entrypoints'
import { DEFAULT_SKIP_VALUES } from '@jiek/pkger/entrypoints'
import { filterLeafs } from '@jiek/pkger/entrypoints'
import { entrypoints2Exports, getAllLeafs, resolveEntrypoints } from '@jiek/pkger/entrypoints'
import { dts } from '@jiek/rollup-plugin-dts'
import { getWorkspaceDir } from '@jiek/utils/getWorkspaceDir'
import json from '@rollup/plugin-json'
import terser from '@rollup/plugin-terser'
import type { OutputOptions, OutputPlugin, Plugin, RollupOptions } from 'rollup'
import esbuild from 'rollup-plugin-esbuild'

export interface TemplateOptions {
}

interface PackageJSON {
  name?: string
  type?: string
  exports?: Record<string, unknown> | string | string[]
}

const {
  JIEK_ROOT
} = process.env
const WORKSPACE_ROOT = JIEK_ROOT ?? getWorkspaceDir()
const COMMON_OPTIONS = {} satisfies RollupOptions
const COMMON_PLUGINS = [
  json()
]

const resolveWorkspacePath = (p: string) => resolve(WORKSPACE_ROOT, p)

const pascalCase = (str: string) =>
  str
    .replace(/[@|/-](\w)/g, (_, $1) => $1.toUpperCase())
    .replace(/(?:^|-)(\w)/g, (_, $1) => $1.toUpperCase())

const reveal = (obj: string | Record<string, unknown>, keys: string[]) =>
  keys.reduce((acc, key) => {
    if (typeof acc === 'string') throw new Error('key not found in exports')
    if (!(key in acc)) throw new Error(`key ${key} not found in exports`)
    return acc[key] as string | Record<string, unknown>
  }, obj)

const withMinify = (
  output: OutputOptions & {
    plugins?: OutputPlugin[]
  }
) => [
  output,
  {
    ...output,
    file: output.file?.replace(/(\.[cm]?js)$/, '.min$1'),
    plugins: [
      ...(output.plugins ?? []),
      terser()
    ]
  }
]

const generateConfigs = ({
  name,
  input,
  output,
  outdir,
  pkgIsModule,
  conditionals
}: {
  name: string
  input: string
  output: string
  outdir: string
  pkgIsModule: boolean
  conditionals: string[]
}) => {
  const isModule = conditionals.includes('import')
  const isCommonJS = conditionals.includes('require')
  const isBrowser = conditionals.includes('browser')
  return [
    {
      input,
      output: [
        ...withMinify({
          file: output,
          name,
          format: isModule ? 'esm' : (
            isCommonJS ? 'cjs' : (
              isBrowser ? 'umd' : (
                pkgIsModule ? 'esm' : 'cjs'
              )
            )
          )
        })
      ],
      plugins: [
        esbuild()
      ]
    },
    {
      input,
      output: [
        { dir: outdir }
      ],
      plugins: [
        dts({ tsconfig: resolveWorkspacePath('tsconfig.dts.json') })
      ]
    }
  ]
}

export function template(packageJSON: PackageJSON, options: TemplateOptions = {}) {
  const { name, type, exports: entrypoints } = packageJSON
  const pkgIsModule = type === 'module'
  const outdir = 'dist'
  if (!name) throw new Error('package.json name is required')
  if (!entrypoints) throw new Error('package.json exports is required')

  const packageName = pascalCase(name)

  const [, resolvedEntrypoints] = resolveEntrypoints(entrypoints)
  const filteredResolvedEntrypoints = filterLeafs(
    resolvedEntrypoints as RecursiveRecord<string>,
    {
      skipValue: [
        // ignore values that filename start with `.jk-noentry`
        /(^|\/)\.jk-noentry/,
        ...DEFAULT_SKIP_VALUES
      ]
    }
  )
  const exports = entrypoints2Exports(filteredResolvedEntrypoints, {})

  const leafMap = new Map<string, string[][]>()
  getAllLeafs(filteredResolvedEntrypoints as RecursiveRecord<string>, ({ keys, value }) => {
    if (typeof value === 'string') {
      const keysArr = leafMap.get(value) ?? []
      leafMap.set(value, keysArr)
      keysArr.push(keys)
    }
    return false
  })

  const configs: (RollupOptions & {
    plugins: Plugin[]
  })[] = []
  leafMap.forEach((keysArr, input) =>
    keysArr.forEach((keys) => {
      const [path, ...conditionals] = keys

      const name = packageName + (path === '.' ? '' : pascalCase(path))
      const keyExports = reveal(exports, keys)

      switch (typeof keyExports) {
        case 'string': {
          configs.push(...generateConfigs({
            name,
            input,
            output: keyExports,
            outdir,
            pkgIsModule,
            conditionals
          }))
          break
        }
        case 'object': {
          getAllLeafs(keyExports as RecursiveRecord<string>, ({ keys: nextKeys, value }) => {
            const allConditionals = [...new Set([...conditionals, ...nextKeys])]
            if (typeof value === 'string') {
              configs.push(...generateConfigs({
                name,
                input,
                output: value,
                outdir,
                pkgIsModule,
                conditionals: allConditionals
              }))
            }
            return false
          })
          break
        }
      }
    })
  )
  return configs.map(c => ({
    ...COMMON_OPTIONS,
    ...c,
    plugins: [
      ...COMMON_PLUGINS,
      ...(c.plugins ?? [])
    ]
  }))
}
