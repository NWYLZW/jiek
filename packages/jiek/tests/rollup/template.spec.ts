import { template } from 'jiek/rollup.v2'
import { describe, expect, test } from 'vitest'

describe('rollup - template', () => {
  test('base', () => {
    expect(template({
      name: 'foo',
      exports: './src/index.ts'
    })).toMatchSnapshot()
    expect(template({
      type: 'module',
      name: 'foo',
      exports: './src/index.ts'
    })).toMatchSnapshot()
  })
  test('with package.json and *.d.ts export files and should skip it', () => {
    expect(template({
      name: 'foo',
      exports: {
        './package.json': './package.json'
      }
    })).toMatchSnapshot()
    expect(template({
      name: 'foo',
      exports: {
        './package.json': './package.json',
        '.': './src/index.ts'
      }
    })).toMatchSnapshot()
    expect(template({
      name: 'foo',
      exports: {
        './package.json': './package.json',
        '.': './src/index.ts',
        './foo': './src/foo.d.ts'
      }
    })).toMatchSnapshot()
  })
  test('with import or require conditional', () => {
    expect(template({
      name: 'foo',
      exports: {
        '.': {
          import: './src/index.mts',
          default: './src/index.ts'
        }
      }
    })).toMatchSnapshot()
    expect(template({
      type: 'module',
      name: 'foo',
      exports: {
        '.': {
          require: './src/index.cts',
          default: './src/index.ts'
        }
      }
    })).toMatchSnapshot()
  })
})
