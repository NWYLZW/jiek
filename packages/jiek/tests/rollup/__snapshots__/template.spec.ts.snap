// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rollup - template > base 1`] = `
[
  {
    "input": "./src/index.ts",
    "output": [
      {
        "file": "./dist/index.js",
        "format": "cjs",
        "name": "Foo",
      },
      {
        "file": "./dist/index.min.js",
        "format": "cjs",
        "name": "Foo",
        "plugins": [
          {
            "name": "terser",
            "numOfWorkersUsed": 0,
            "renderChunk": [Function],
          },
        ],
      },
    ],
    "plugins": [
      {
        "name": "json",
        "transform": [Function],
      },
      {
        "buildStart": [Function],
        "name": "esbuild",
        "options": [Function],
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
    ],
  },
  {
    "input": "./src/index.ts",
    "output": [
      {
        "dir": "dist",
      },
    ],
    "plugins": [
      {
        "name": "json",
        "transform": [Function],
      },
      {
        "name": "dts",
        "options": [Function],
        "outputOptions": [Function],
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
    ],
  },
]
`;

exports[`rollup - template > base 2`] = `
[
  {
    "input": "./src/index.ts",
    "output": [
      {
        "file": "./dist/index.js",
        "format": "esm",
        "name": "Foo",
      },
      {
        "file": "./dist/index.min.js",
        "format": "esm",
        "name": "Foo",
        "plugins": [
          {
            "name": "terser",
            "numOfWorkersUsed": 0,
            "renderChunk": [Function],
          },
        ],
      },
    ],
    "plugins": [
      {
        "name": "json",
        "transform": [Function],
      },
      {
        "buildStart": [Function],
        "name": "esbuild",
        "options": [Function],
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
    ],
  },
  {
    "input": "./src/index.ts",
    "output": [
      {
        "dir": "dist",
      },
    ],
    "plugins": [
      {
        "name": "json",
        "transform": [Function],
      },
      {
        "name": "dts",
        "options": [Function],
        "outputOptions": [Function],
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
    ],
  },
]
`;

exports[`rollup - template > with import or require conditional 1`] = `
[
  {
    "input": "./src/index.mts",
    "output": [
      {
        "file": "./dist/index.mjs",
        "format": "esm",
        "name": "Foo",
      },
      {
        "file": "./dist/index.min.mjs",
        "format": "esm",
        "name": "Foo",
        "plugins": [
          {
            "name": "terser",
            "numOfWorkersUsed": 0,
            "renderChunk": [Function],
          },
        ],
      },
    ],
    "plugins": [
      {
        "name": "json",
        "transform": [Function],
      },
      {
        "buildStart": [Function],
        "name": "esbuild",
        "options": [Function],
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
    ],
  },
  {
    "input": "./src/index.mts",
    "output": [
      {
        "dir": "dist",
        "entryFileNames": [Function],
      },
    ],
    "plugins": [
      {
        "name": "json",
        "transform": [Function],
      },
      {
        "name": "dts",
        "options": [Function],
        "outputOptions": [Function],
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
    ],
  },
  {
    "input": "./src/index.ts",
    "output": [
      {
        "file": "./dist/index.js",
        "format": "cjs",
        "name": "Foo",
      },
      {
        "file": "./dist/index.min.js",
        "format": "cjs",
        "name": "Foo",
        "plugins": [
          {
            "name": "terser",
            "numOfWorkersUsed": 0,
            "renderChunk": [Function],
          },
        ],
      },
    ],
    "plugins": [
      {
        "name": "json",
        "transform": [Function],
      },
      {
        "buildStart": [Function],
        "name": "esbuild",
        "options": [Function],
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
    ],
  },
  {
    "input": "./src/index.ts",
    "output": [
      {
        "dir": "dist",
        "entryFileNames": [Function],
      },
    ],
    "plugins": [
      {
        "name": "json",
        "transform": [Function],
      },
      {
        "name": "dts",
        "options": [Function],
        "outputOptions": [Function],
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
    ],
  },
]
`;

exports[`rollup - template > with import or require conditional 2`] = `
[
  {
    "input": "./src/index.cts",
    "output": [
      {
        "file": "./dist/index.cjs",
        "format": "cjs",
        "name": "Foo",
      },
      {
        "file": "./dist/index.min.cjs",
        "format": "cjs",
        "name": "Foo",
        "plugins": [
          {
            "name": "terser",
            "numOfWorkersUsed": 0,
            "renderChunk": [Function],
          },
        ],
      },
    ],
    "plugins": [
      {
        "name": "json",
        "transform": [Function],
      },
      {
        "buildStart": [Function],
        "name": "esbuild",
        "options": [Function],
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
    ],
  },
  {
    "input": "./src/index.cts",
    "output": [
      {
        "dir": "dist",
        "entryFileNames": [Function],
      },
    ],
    "plugins": [
      {
        "name": "json",
        "transform": [Function],
      },
      {
        "name": "dts",
        "options": [Function],
        "outputOptions": [Function],
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
    ],
  },
  {
    "input": "./src/index.ts",
    "output": [
      {
        "file": "./dist/index.js",
        "format": "esm",
        "name": "Foo",
      },
      {
        "file": "./dist/index.min.js",
        "format": "esm",
        "name": "Foo",
        "plugins": [
          {
            "name": "terser",
            "numOfWorkersUsed": 0,
            "renderChunk": [Function],
          },
        ],
      },
    ],
    "plugins": [
      {
        "name": "json",
        "transform": [Function],
      },
      {
        "buildStart": [Function],
        "name": "esbuild",
        "options": [Function],
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
    ],
  },
  {
    "input": "./src/index.ts",
    "output": [
      {
        "dir": "dist",
        "entryFileNames": [Function],
      },
    ],
    "plugins": [
      {
        "name": "json",
        "transform": [Function],
      },
      {
        "name": "dts",
        "options": [Function],
        "outputOptions": [Function],
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
    ],
  },
]
`;

exports[`rollup - template > with package.json and *.d.ts export files and should skip it 1`] = `[]`;

exports[`rollup - template > with package.json and *.d.ts export files and should skip it 2`] = `
[
  {
    "input": "./src/index.ts",
    "output": [
      {
        "file": "./dist/index.js",
        "format": "cjs",
        "name": "Foo",
      },
      {
        "file": "./dist/index.min.js",
        "format": "cjs",
        "name": "Foo",
        "plugins": [
          {
            "name": "terser",
            "numOfWorkersUsed": 0,
            "renderChunk": [Function],
          },
        ],
      },
    ],
    "plugins": [
      {
        "name": "json",
        "transform": [Function],
      },
      {
        "buildStart": [Function],
        "name": "esbuild",
        "options": [Function],
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
    ],
  },
  {
    "input": "./src/index.ts",
    "output": [
      {
        "dir": "dist",
      },
    ],
    "plugins": [
      {
        "name": "json",
        "transform": [Function],
      },
      {
        "name": "dts",
        "options": [Function],
        "outputOptions": [Function],
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
    ],
  },
]
`;

exports[`rollup - template > with package.json and *.d.ts export files and should skip it 3`] = `
[
  {
    "input": "./src/index.ts",
    "output": [
      {
        "file": "./dist/index.js",
        "format": "cjs",
        "name": "Foo",
      },
      {
        "file": "./dist/index.min.js",
        "format": "cjs",
        "name": "Foo",
        "plugins": [
          {
            "name": "terser",
            "numOfWorkersUsed": 0,
            "renderChunk": [Function],
          },
        ],
      },
    ],
    "plugins": [
      {
        "name": "json",
        "transform": [Function],
      },
      {
        "buildStart": [Function],
        "name": "esbuild",
        "options": [Function],
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
    ],
  },
  {
    "input": "./src/index.ts",
    "output": [
      {
        "dir": "dist",
      },
    ],
    "plugins": [
      {
        "name": "json",
        "transform": [Function],
      },
      {
        "name": "dts",
        "options": [Function],
        "outputOptions": [Function],
        "renderChunk": [Function],
        "resolveId": [Function],
        "transform": [Function],
      },
    ],
  },
]
`;
